src/sensors/trilateration.cpp

```diff
@@ -0,0 +1,368 @@
+/**
+ * Hydrophone Trilateration (Position Derivation) Node
+ *
+ * @author Ryan Summers
+ * @date 12-12-2016
+ *
+ * @brief This file contains the calculations of the submarine's position
+ *        relative to the pinger from a number of hydrophone time stamps.
+ *
+ * @note The mathematic equations presented in this file are derived at the
+ *       following two links. Please refer to them for all questions relating
+ *       to mathematic operations. All variables have been named after the
+ *       variables defined in the derivation for clarity.
+ *          http://robosub.eecs.wsu.edu/wiki/ee/hydrophones/trilateration/start
+ *          http://robosub.eecs.wsu.edu/wiki/cs/hydrophones/start
+ */
+
+#include "robosub/HydrophoneDeltas.h"
+#include "robosub/PositionsStamped.h"
+#include "robosub/QuaternionStampedAccuracy.h"
+#include "robosub/depth_stamped.h"
+#include "geometry_msgs/Vector3.h"
+#include "ros/ros.h"
+#include <eigen3/Eigen/Geometry>
+#include <eigen3/Eigen/Dense>
+
+#include <vector>
+
+using std::vector;
+using Eigen::Vector3d;
+using Eigen::Quaterniond;
+
+ros::Publisher position_pub;
+ros::Publisher position_all_pub;
+
+/*
+ * The speed of sound in water is 1484 m/s.
+ */
+static constexpr double speed_sound_in_water = 1484;
+
+/*
+ * The last read submarine orientation and depth information.
+ */
+Quaterniond current_orientation;
+float current_depth;
+
+/*
+ * The depth of the pinger in the pool.
+ */
+float pinger_depth;
+
+/*
+ * The reference hydrophone position with respect to the center of the
+ * submarine.
+ */
+Vector3d reference_hydrophone_position;
+
+/*
+ * The hydrophone position arguments. Delta is the position of the hx
+ * hydrophone along the x axis, epsilon is hy along the y axis, and zeta is hz
+ * along the z axis.
+ */
+double delta;
+double epsilon;
+double zeta;
+
+/**
+ * Callback function to receive the submarine's current orientation.
+ *
+ * @param msg A pointer to the received message.
+ *
+ * @return None.
+ */
+void orientationCallback(const
+        robosub::QuaternionStampedAccuracy::ConstPtr& msg)
+{
+    current_orientation = Quaterniond(msg->quaternion.w, msg->quaternion.x,
+            msg->quaternion.y, msg->quaternion.z);
+}
+
+/**
+ * Callback function to receive the submarine's current depth.
+ *
+ * @param msg A pointer to the received message.\
+ *
+ * @return None.
+ */
+void depthCallback(const robosub::depth_stamped::ConstPtr& msg)
+{
+    current_depth = msg->depth;
```

Comment:
If all of the depth sensors will be read by the same node, you should be able to use the same system as the position messages and vision messages to allow this to be scalable.